### ISSUE TITLE
Warn when match variables start with an uppercase letter

### ISSUE THREAD

**Issue by kmcallister** [rust-lang/rust#10304]
I had code like

~~~ .rs
match e.kind {
    EndOfFile => return Ok(()),
    _         => return Err(()),
}
~~~

and got `error: unreachable pattern` on the second pattern.  I hadn't imported `std::rt::io::EndOfFile` so `rustc` interpreted `EndOfFile` as a variable rather than an enum constructor.

Most Rust code follows the stylistic rule of capitalizing enum constructors and not capitalizing variables.  So I think `rustc` should warn when binding a capitalized variable in a `match` — at least when it makes another pattern unreachable, but probably always.

**Comment by steveklabnik** [comment]
Triage: still no lint, still same error message.

**Comment by wesleywiser** [comment]
Is this still an issue? I tried the repro from #13995 and I got this

```
test.rs:9:9: 9:11 warning: pattern binding `V1` is named the same as one of the variants of the type `A::E` [E0170]
test.rs:9         V1 => 1,
                  ^~
test.rs:9:9: 9:11 help: run `rustc --explain E0170` to see a detailed explanation
test.rs:9:9: 9:11 help: if you meant to match on a variant, consider making the path in the pattern qualified: `A::E::V1`
test.rs:10:9: 10:11 warning: pattern binding `V2` is named the same as one of the variants of the type `A::E` [E0170]
test.rs:10         V2 => 2,
                   ^~
test.rs:10:9: 10:11 help: run `rustc --explain E0170` to see a detailed explanation
test.rs:10:9: 10:11 help: if you meant to match on a variant, consider making the path in the pattern qualified: `A::E::V2`
test.rs:10:9: 10:11 error: unreachable pattern [E0001]
test.rs:10         V2 => 2,
                   ^~
test.rs:10:9: 10:11 help: run `rustc --explain E0001` to see a detailed explanation
error: aborting due to previous error
```

**Comment by alexcrichton** [comment]
I also believe this is fixed, thanks @wesleywiser!
